terraform/lib/dependabot/terraform/registry_client.rb

```diff
@@ -98,6 +80,43 @@ def headers_for(hostname)
         token = tokens[hostname]
         token ? { "Authorization" => "Bearer #{token}" } : {}
       end
+
+      def services
+        @services ||=
+          begin
+            response = http_get("https://#{hostname}/.well-known/terraform.json")
+            response.status == 200 ? JSON.parse(response.body) : {}
+          end
+      end
+
+      def service_key_for(name)
+        case name
+        when "module", "modules", "registry"
+          "modules.v1"
+        when "provider", "providers"
+          "providers.v1"
+        else
+          raise "Invalid source type"
+        end
+      end
+
+      def service_url_for(service_key)
+        "https://#{hostname}#{services.fetch(service_key)}"
+      rescue KeyError
+        raise "Host does not support required Terraform-native service"
```

Comment:
> then discovery fails and Terraform considers the host to not support any Terraform-native services.
https://www.terraform.io/docs/internals/remote-service-discovery.html#discovery-process

Reply:
I wonder if this should be something other than a `RuntimeError`, maybe a [`Dependabot::PrivateSourceAuthenticationError`](https://github.com/dependabot/dependabot-core/blob/955eccebe71140cf00d5fde3e6b5a8cb2ff3be70/common/lib/dependabot/errors.rb#L122) or something other than inherits from `Dependabot::Error` that we know how to handle. This would currently show up as an "unknown error" in ours logs/metrics etc

Reply:
> I wonder if this should be something other than a `RuntimeError`, maybe a [`Dependabot::PrivateSourceAuthenticationError`](https://github.com/dependabot/dependabot-core/blob/955eccebe71140cf00d5fde3e6b5a8cb2ff3be70/common/lib/dependabot/errors.rb#L122) or something other than inherits from `Dependabot::Error` that we know how to handle. This would currently show up as an "unknown error" in ours logs/metrics etc

Good call. I'll update to make sure this doesn't effect the SLO's.

Reply:
Any reason this didn't become an `Dependabot::PrivateSourceAuthenticationError`? It seems more descriptive to me, but maybe I'm misunderstanding how we use that error?